### PR TITLE
Update implementation to meet agreed requirements in 2022-06-23 meeting

### LINKED ISSUE #123: Update implementation to meet agreed requirements in 2022-06-23 meeting
This is new functionality. We will need to re-build the terms and definitions using a combination of XML and EXPRESS code.

1. We only keep ENTITY definitions in ISO 10303-2. Remove the TYPE, FUNCTION, RULE, constant definitions/descriptions from the document. This information is only provided in the EXPRESS code. 

2. As every Clause 4/5 term entry has a source, remove the NOTE, and add the “incompletely defined in this document” sentence to Clause 4 and 5 at the top.

3. Alter definition format to use a generated sentence directly from EXPRESS, keep `descriptions.xml` text (the current text we have) as a Note to entry of the term.

In order to achieve this, we have to do this.
Today, we only have the left flow. We need to add the right flow using Expressir.

Screenshot 2022-06-30 at 7 47 34 PM

This is what we need to do:
Screen Shot 2022-08-09 at 12 49 45 AM

## Resulting structure of a term

* the "term name" is the entity name, which comes from the `{schema-name}/{schema-name}.exp` file
* the "definition" is generated from EXPRESS information of that entity from the `{schema-name}/{schema-name}.exp` file (schema name => domain, entity type => description...)
* the "Note 1" is generated from a lookup of the entity name in the currently parsed database of `{schema-name}/descriptions.xml`

## Flow

Sample code:
```sh
$ gem install expressir
$ irb
> require ‘expressir'
> require ‘expressir/express/parser'
> exp_file = ‘my-file-name.exp'
> repo = Expressir::Express::Parser.from_file(exp_file)
```

This is only for the "Clause 4 and 5 terms", i.e. the files that get generated are the `04x-*.adoc` and `05x-*.adoc` files.

What this means:
1. For all data in a chosen directory, read all the `*.exp` files. This is achieved by using the `Expressir::Express::Repo` class (something like that), it will provide you with an object (the parse tree) that contains all the EXPRESS schemas of a given directory.
2. For every `ENTITY` object in the loaded schemas, generate this output (which is built using information from `{schema-name}/{schema-name}.exp` instead of the current definition (which is sourced from `{schema-name}/descriptions.xml`. Move the current definition into a `NOTE:` block.

## Types of output

There are 3 kinds of ENTITY input with different outputs:
1. Single SUBTYPE
2. Multiple SUBTYPEs
3. No SUBTYPE

For the `{"resource" or "module"}` value:
* if the file path is `resources/*/{schema-name}.exp`, it is a "resource"
* If the file path is `modules/*/{schema-name}.exp`, it is a "module"

### Single SUBTYPE

EXPRESS:
```
ENTITY geometric_representation_context
  SUBTYPE OF (representation_context);
  coordinate_space_dimension : dimension_count;
END_ENTITY;
```

Output: (adoc, the curly braces indicate variables)
```adoc
=== {ENTITY NAME}
domain:[{"resource" or "module"}: {SCHEMA NAME}]

entity data type that is a type of {SUBTYPE} that represents a/an {ENTITY NAME} entity

NOTE: {ORIGINAL DEFINITION}
```

Example:
```adoc
=== geometric_representation_context
domain:[resource: product_definition_schema]

entity data type that is a type of representation_context that represents a geometric_representation_context entity

NOTE: A geometric_representation_context is a type of representation_context in which geometric_representation_item s are geometrically founded.
```

### Multiple SUBTYPEs

EXPRESS:
```
ENTITY characterized_product
    SUBTYPE OF (characterized_object, product);
END_ENTITY;
```

Output: (adoc, the curly braces indicate variables)
```adoc
=== {ENTITY NAME}
domain:[{"resource" or "module"}: {SCHEMA NAME}]

entity data type that is a type of {SUBTYPE 1} and {SUBTYPE 2...} that represents a/an {ENTITY NAME} entity

NOTE: {ORIGINAL DEFINITION}
```

Example:
```adoc
=== characterized_product
domain:[resource: product_definition_schema]

entity data type that is a type of characterized_object and product that represents a characterized_product entity

NOTE: A characterized_product is a type of characterized_object and a type of product for the purpose of characterizing a product.
```

### No SUBTYPE

EXPRESS:
```
ENTITY product
  ...
END_ENTITY;
```

Output: (adoc, the curly braces indicate variables)
```adoc
=== {ENTITY NAME}
domain:[{"resource" or "module"}: {SCHEMA NAME}]

entity data type that represents a/an {ENTITY NAME} entity

NOTE: {ORIGINAL DEFINITION}
```

Example:
```adoc
=== product
domain:[resource: product_definition_schema]

entity data type that represents a product entity

NOTE: thing or substance produced by a natural or artificial process.
```

### PR DESCRIPTION
**Pull request by HassanAkbar**
### Metanorma PR checklist

 - [ ] We only keep ENTITY definitions in ISO 10303-2. Remove the TYPE, FUNCTION, RULE, constant definitions/descriptions from the document. This information is only provided in the EXPRESS code.

 - [x] As every Clause 4/5 term entry has a source, remove the NOTE, and add the “incompletely defined in this document” sentence to Clause 4 and 5 at the top.

 - [x] Alter definition format to use a generated sentence directly from EXPRESS, keep descriptions.xml text (the current text we have) as a Note to entry of the term.

 - [x] Add new Test cases and fix failing ones.

### Example Output

```adoc

== Surface_conditions_arm

This term is incompletely defined in this document.
// STEPmod path: modules/surface_conditions/arm_descriptions.xml

=== sc_additional_application_domain_enumeration

domain:[module: Surface_conditions_arm]

entity data type that represents a/an sc_additional_application_domain_enumeration entity

NOTE: The *sc_additional_application_domain_enumeration* provides predefined identifiers for the additional application domains of *Product_view_definition* s.


[.source]
<<ISO-TS_10303-1110_2021>>


// STEPmod path: modules/surface_conditions/arm_descriptions.xml

=== surface_lay_and_orientation

domain:[module: Surface_conditions_arm]

entity data type that represents a/an surface_lay_and_orientation entity

NOTE: The *surface_lay_and_orientation* provides predefined identifiers of the surface lay as specified in table 2 of ISO 1302:2002.


[.source]
<<ISO-TS_10303-1110_2021>>

...
```
 
closes #123

### CODE REVIEW

**Comment by HassanAkbar** [comment]
@ronaldtse  I have a question related to this, currently we are iterating on the `ext_description` in `descriptions.xml` file and then finding the related `ENTITY` in the `.exp` file to generate the `description`, but the issue is there are some cases where there is an `ext_description` present but not the coresponding `ENTITY` (e.g `restrict_support_structure_category` has `ext_description` inside `modules/additive_manufacturing_part_and_build_information/mim_descriptions.xml` but there is no `ENTITY Additive_manufacturing_part_and_build_information`), so in that case I am using the `No SUBTYPE` template for definition.

Is this approach correct or now after switching to `.exp` files we should iterate on `ENTITY` and find the coresponding `ext_description` for the the `note`?

**Comment by TRThurman** [comment]
As noted there can be extra information in the description xml file.
So we must iterate over the EXPRESS and then find the description.
Once the algorithm is updated, does the follow-on question still exist?

File: exe/stepmod-extract-terms [hunk 1]
```diff
@@ -136,6 +136,7 @@ part_resources.each do |(bibdata, current_part_resources)|
   fn = "04x-stepmod-entities-resources-#{bibdata.part}.adoc"
   File.open(fn, "w") do |file|
     # file.puts("== #{part_to_title(bibdata)}\n\n")
+    file.puts("This term is incompletely defined in this document.\n")
```
Comment (ronaldtse) [inline exe/stepmod-extract-terms:139]: Remove this line -- not necessary.

File: exe/stepmod-extract-terms [hunk 2]
```diff
@@ -164,6 +165,7 @@ part_modules.each do |(bibdata, part_modules_arm, part_modules_mim)|
 
       # puts "SCHEMA NAME ARM: #{schema_name}"
       file.puts("== #{schema_name}\n\n")
+      file.puts("This term is incompletely defined in this document.\n")
```
Comment (ronaldtse) [inline exe/stepmod-extract-terms:168]: Remove, this line will break the document schema.

File: lib/stepmod/utils/concept.rb [hunk 3]
```diff
@@ -11,25 +13,26 @@ class Concept < Glossarist::LocalizedConcept
...
+              exp_repo: exp_repo,
             },
           )
 
           return nil if converted_definition.nil? || converted_definition.strip.empty?
 
           if definition_xml.name == "ext_description"
-            converted_definition = <<~TEXT
-              #{converted_definition}
+            # converted_definition = <<~TEXT
```
Comment (ronaldtse) [inline lib/stepmod/utils/concept.rb:30]: The `converted_definition` should become this:

```adoc
NOTE: #{converted_definition}

NOTE: This term is incompletely defined in this document.
Reference <<#{reference_anchor}>> for the complete definition.
```

**Comment by TRThurman** [comment]
clarifications:
1- in the domain where it is a 'module', the correct term is 'application module'.
2- document source is to be converted to a comment and retained in the asciidoc file.
This:
[.source]
<<ISO-TS_10303-1110_2021>>

becomes this:

//[.source]
//<<ISO-TS_10303-1110_2021>>

**Comment by ronaldtse** [comment]
@HassanAkbar as mentioned by @TRThurman, let's reverse the iteration so that:
1. Iterate over the EXPRESS schema (`.exp`), for all ENTITYs
2. For every ENTITY, look up the corresponding `<ext_description>` in the `descriptions.xml` file
3. Render the term/definition.

**Review by ronaldtse — CHANGES_REQUESTED** [review]

**Comment by TRThurman** [comment]
@ronaldtse, thanks! My comment on commenting out the [.source] is incorrect for clauses 4, 5.

Sorry for confusion.

**Comment by ronaldtse** [comment]
@TRThurman but your message did remind me about commenting out for Clause 3 -- I haven't actually updated stepmod-utils to do this but have done the commenting out manually. Now that we have finalized that Clause 3 terms belongs to the document, we will do the commenting out.

File: lib/stepmod/utils/concept.rb (hunk 3 above)
Comment (ronaldtse) [reply lib/stepmod/utils/concept.rb]: Thank you!

File: lib/stepmod/utils/converters/stepmod_ext_description.rb [hunk 4]
```diff
@@ -43,21 +43,46 @@ def convert(node, state = {})
...
           TEMPLATE
         end
+
+        private
+
+        def extract_entity_text(entity, entity_name)
+          if entity.nil?
+            "entity data type that represents a/an #{entity_name} entity"
+          elsif entity.subtype_of.size.zero?
+            "entity data type that represents a/an #{entity.id} entity"
```
Comment (ronaldtse) [inline lib/stepmod/utils/converters/stepmod_ext_description.rb:77]: We will have to select "an" or "a" using some detection mechanism.

File: lib/stepmod/utils/converters/stepmod_ext_description.rb [hunk 5]
```diff
@@ -43,21 +43,46 @@ def convert(node, state = {})
...
+          elsif entity.subtype_of.size == 1
+            "entity data type that is a type of #{entity.subtype_of[0].id} that represents a/an #{entity.id} entity"
+          else
+            "entity data type that is a type of #{entity.subtype_of.map(&:id).join(' and ')} that represents a/an #{entity.id} entity"
```
Comment (ronaldtse) [inline lib/stepmod/utils/converters/stepmod_ext_description.rb:81]: As above re: a, an.

File: lib/stepmod/utils/converters/stepmod_ext_description.rb [hunk 6]
```diff
@@ -43,21 +43,46 @@ def convert(node, state = {})
...
+
+        private
+
+        def extract_entity_text(entity, entity_name)
+          if entity.nil?
+            "entity data type that represents a/an #{entity_name} entity"
+          elsif entity.subtype_of.size.zero?
+            "entity data type that represents a/an #{entity.id} entity"
+          elsif entity.subtype_of.size == 1
+            "entity data type that is a type of #{entity.subtype_of[0].id} that represents a/an #{entity.id} entity"
```
Comment (ronaldtse) [inline lib/stepmod/utils/converters/stepmod_ext_description.rb:79]: As above re: a, an.

File: lib/stepmod/utils/converters/stepmod_ext_description.rb [hunk 7]
```diff
@@ -43,21 +43,46 @@ def convert(node, state = {})
...
+
+            NOTE: #{child_text}
           TEMPLATE
         end
+
+        private
+
+        def extract_entity_text(entity, entity_name)
+          if entity.nil?
+            "entity data type that represents a/an #{entity_name} entity"
```
Comment (ronaldtse) [inline lib/stepmod/utils/converters/stepmod_ext_description.rb:75]: Need to address "a", "an".

File: spec/stepmod/concept_spec.rb [hunk 8]
```diff
@@ -85,9 +87,11 @@
         // STEPmod path:
         === complement
 
-        domain:[ISO 10303 resource]
+        domain:[resource: action_schema]
+
+        entity data type that represents a/an complement entity
 
-        A *complement* is a relationship that is between
+        NOTE: A *complement* is a relationship that is between
```
Comment (ronaldtse) [inline spec/stepmod/concept_spec.rb:94]: This NOTE is a long one that contains blank lines. Blank lines causes Metanorma to split text into multiple paragraphs. In this case, we need to use an alternative syntax of NOTE.

The following three forms are equivalent:

Single paragraph:
```adoc
NOTE: This is a note.
```

Continuation style:
```adoc
NOTE: This is a note.
+
* list in a note.
```

Block style:
```adoc
[NOTE]
--
This is a note.

* list in a note.
--
```

We need to use the block-style syntax (the third one).

File: spec/stepmod/terms_extractor_spec.rb [hunk 9]
```diff
@@ -40,39 +40,33 @@
...
-        domain:[ISO 10303 application module]
-
-        An *applied_action_assignment* is an <<express:action_schema.action,action>> related to the data that are affected by the <<express:action_schema.action,action>>. An *applied_action_assignment* is a type of <<express:management_resources_schema.action_assignment,action_assignment>>.
+        domain:[module: Activity_mim]
 
+        entity data type that is a type of action_assignment that represents a/an applied_action_assignment entity
 
-        NOTE: This term is incompletely defined in this document.
-        Reference <<ISO-TS_10303-1047_2014>> for the complete definition.
+        NOTE: An *applied_action_assignment* is an <<express:action_schema.action,action>> related to the data that are affected by the <<express:action_schema.action,action>>. An *applied_action_assignment* is a type of <<express:management_resources_schema.action_assignment,action_assignment>>.
```
Comment (ronaldtse) [inline spec/stepmod/terms_extractor_spec.rb:69]: We need to add one additional NOTE (a second NOTE) to say that:

```adoc
NOTE: This term is incompletely defined in this document. Reference <<ISO-TS_10303-1047_2014>> for the complete definition.
```

File: spec/stepmod/terms_extractor_spec.rb [hunk 10]
```diff
@@ -40,39 +40,33 @@
...
-        domain:[ISO 10303 application module]
-
-        An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute.
+        domain:[module: Activity_arm]
 
+        entity data type that represents a/an Activity entity
 
-        NOTE: This term is incompletely defined in this document.
-        Reference <<ISO-TS_10303-1047_2014>> for the complete definition.
+        NOTE: An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute.
```
Comment (ronaldtse) [inline spec/stepmod/terms_extractor_spec.rb:58]: Need to add an additional NOTE below this one:
```adoc
NOTE: This term is incompletely defined in this document. Reference <<ISO-TS_10303-1047_2014>> for the complete definition.
```

File: spec/stepmod/terms_extractor_spec.rb [hunk 11]
```diff
@@ -40,39 +40,33 @@
...
-        domain:[ISO 10303 resource]
+        domain:[resource: action_schema]
 
-        An *action_directive* is an authoritative instrument that provides directions to achieve the specified results.
+        entity data type that represents a/an action_directive entity
 
-
-        NOTE: This term is incompletely defined in this document.
-        Reference <<ISO_10303-41_2021>> for the complete definition.
+        NOTE: An *action_directive* is an authoritative instrument that provides directions to achieve the specified results.
```
Comment (ronaldtse) [inline spec/stepmod/terms_extractor_spec.rb:47]: See below, need to add the 'incompletely defined' NOTE as a second NOTE.

File: spec/stepmod/concept_spec.rb [hunk 12]
```diff
@@ -96,10 +100,6 @@
         that indicates set S2 consists of all members of U that are not members of S1.
 
 
-        NOTE: This term is incompletely defined in this document.
```
Comment (ronaldtse) [inline spec/stepmod/concept_spec.rb:99]: We still need to keep this NOTE for items in Clause 4 (`04x-*`) and Clause 5 (`05x-*.adoc`).

Comment (HassanAkbar) [reply spec/stepmod/concept_spec.rb:99]: > 3. As every Clause 4/5 term entry has a source, remove the NOTE, and add the “incompletely defined in this document” sentence to Clause 4 and 5 at the top.

@ronaldtse this is mentioned in the [ticket](https://github.com/metanorma/stepmod-utils/issues/123), are we not doing this right now or am I missing something here?

Comment (ronaldtse) [reply spec/stepmod/concept_spec.rb:99]: Yes you are correct. I got confused. Instead please ensure we have the “[.source]” entry for the Clause 4 and Clause 5. Thanks.

**Comment by HassanAkbar** [comment]
@ronaldtse there are some entities in `.exp` files that do not have a corresponding `ext_description` in `.xml` file, e.g [`modules/cardinality_of_relationship`](https://github.com/metanorma/iso-10303-stepmod-wg12/tree/de05521fedcadb45f7d86431bb614eb6b8b71d60/data/modules/cardinality_of_relationship) does not have a `arm_descriptions.xml` file which contains `ext_descriptions` for `Cardinality_of_relationship_arm` schema.

What should be the note in those cases?

**Comment by TRThurman** [comment]
This is a case where the descriptions are in arm.xml.
There are also descriptions in mim.xml

**Comment by HassanAkbar** [comment]
In the existing code we were only looking in `arm_descriptions.xml` and `mim_desctiptions.xml` both of which are not present in case of [`modules/cardinality_of_relationship`](https://github.com/metanorma/iso-10303-stepmod-wg12/tree/de05521fedcadb45f7d86431bb614eb6b8b71d60/data/modules/cardinality_of_relationship). However `arm.xml` and `mim.xml` files are present. If we want to use these then the flow will be something like below

**For `module` folder**
1. Loop on entities in `arm.exp` file
2. Find the corresponding entity in `arm_description.xml` file if it exists.
3. If `arm_description.xml` file does not exist then look for definition in `arm.xml` file
4. Loop through entities in `mim.exp` file that are not found in the previous loop
5. Find the corresponding entity in `mim_description.xml` file if it exists.
6. If `mim_description.xml` file does not exist then look for definition in `mim.xml` file

Is this flow correct?

**Comment by TRThurman** [comment]
Yes

**Comment by ronaldtse** [comment]
@HassanAkbar actually there is an issue when we work with the XML files. Somehow when I issued the instructions, I forgot about https://github.com/metanorma/iso-10303-2/issues/94.

The goal here is actually **NOT** to work with XML files. The goal here is to directly read in [Annotated EXPRESS](https://github.com/metanorma/annotated-express/) files.

For "Resources":
* e.g. for `resources/action_schema`, 
    * Original: https://github.com/metanorma/iso-10303-stepmod/blob/main/data/resources/action_schema/action_schema.exp
    * Annotated: https://github.com/metanorma/iso-10303-stepmod/blob/main/data/resources/action_schema/action_schema_annotated.exp
* The Annotated EXPRESS files already incorporate all the content from `descriptions.xml`
* This means that by using Expressir you can already obtain all the information.

For "Modules":
* e.g. for `modules/cardinality_of_relationship`,
    * Original:
      * https://github.com/metanorma/iso-10303-stepmod/blob/main/data/modules/cardinality_of_relationship/arm.exp
      * https://github.com/metanorma/iso-10303-stepmod/blob/main/data/modules/cardinality_of_relationship/mim.exp
    * Annotated:
      * MISSING! ( to be implemented here: https://github.com/metanorma/stepmod-utils/blob/main/exe/stepmod-annotate-all)
* The Annotated EXPRESS files *should* already incorporate all the content from `arm.xml` and `mim.xml`.

Here's how the flow works:
1. We pull the STEPmod repository (iso-10303-stepmod-wg12) using this workflow https://github.com/metanorma/iso-10303-stepmod-migration/blob/main/.github/workflows/annotate.yml
2. The workflow executes stepmod-utils (this gem) and runs `stepmod-annotate-all` to merge all *.xml content into Annotated EXPRESS files (.exp files but with structured comments)
3. The workflow pushes the repository with the added annotations back

So the task here @HassanAkbar is actually:
1. For Resources: change the code here to read from the resources EXPRESS files
2. For Modules: update `stepmod-annotate-all` to also cover the conversion of Modules to Annotated EXPRESS
3. For Modules: change the code here to read from the modules EXPRESS files

**Comment by HassanAkbar** [comment]
@ronaldtse The definition for the activity in the annotated file is
```
An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute.
[example]
====
Change, distilling, design, a process to drill a hole, and a task such as training someone, are examples of activities.
====

[NOTE]
--
Status information identifying the level of completion of each activity may be provided within an instance of <<express:Activity_arm.Activity_status,Activity_status>>.
--

[NOTE]
--
The items that are affected by an *Activity*, for example as input or output, may be identified within an instance of <<express:Activity_arm.Applied_activity_assignment,Applied_activity_assignment>>.
--
```

As you can see it includes `notes` and `examples`. When I add this in the new definition as a `note` the existing `notes` the formatting is distorted. How do you suggest we resolve this?

**Comment by ronaldtse** [comment]
You need a line break after the definition paragraph and `[example]`.

**Comment by HassanAkbar** [comment]
@ronaldtse this is the old definition when I add it as a note to the new definition like this

```
```

**Comment by ronaldtse** [comment]
We have to fix the format. The correct representation is below:

```adoc
=== Activity
domain:[application module: Activity_arm]

entity data type that represents an Activity entity

[example]
====
Change, distilling, design, a process to drill a hole, and a task such as training someone, are examples of activities.
====

[NOTE]
--
An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute.
--

[NOTE]
--
Status information identifying the level of completion of each activity may be provided within an instance of <<express:Activity_arm.Activity_status,Activity_status>>.
--

[NOTE]
--
The items that are affected by an *Activity*, for example as input or output, may be identified within an instance of <<express:Activity_arm.Applied_activity_assignment,Applied_activity_assignment>>.
--
```

**Comment by HassanAkbar** [comment]
@ronaldtse I think I am not able to explain the problem clearly. The flow that we need to follow is

- Read from annotated EXPRESS file using `expressir` and create a definition.

The annotated EXPRESS file for `activity`(arm_annotated.exp) contains 

```
(*
$Id: arm.exp,v 1.18 2009/12/11 13:39:31 mikeward Exp $
ISO TC184/SC4/WG12 N2941 - ISO/TS 10303-1047 Activity - EXPRESS ARM
*) 


SCHEMA Activity_arm;

USE FROM Activity_method_arm;    -- ISO/TS 10303-1049


TYPE activity_item = EXTENSIBLE GENERIC_ENTITY SELECT;
END_TYPE; 

ENTITY Activity;
  id : STRING;
  name : STRING;
  description : OPTIONAL STRING;
  chosen_method : Activity_method;
END_ENTITY;

...

END_SCHEMA;  -- Activity_arm


(*"Activity_arm.Activity"
An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute. 
[example]
====
Change, distilling, design, a process to drill a hole, and a task such as training someone, are examples of activities.
====

[NOTE]
--
Status information identifying the level of completion of each activity may be provided within an instance of <<express:Activity_arm.Activity_status,Activity_status>>.
--

[NOTE]
--
The items that are affected by an *Activity*, for example as input or output, may be identified within an instance of <<express:Activity_arm.Applied_activity_assignment,Applied_activity_assignment>>.
--
*)

(*"Activity_arm.Activity.id"
the identifier that distinguishes the *Activity*.
*)

(*"Activity_arm.Activity.name"
the words by which the *Activity* is known.
*)

...

```

The issue is when I read the old definition it contains `examples` and `notes`. 
Removing the `notes` and `examples` from the `old definition` and adding them to the new definition will require some string manipulations or again reading the `xml` files because currently, I didn't find a way to extract these from the express file.

**Comment by ronaldtse** [comment]
In today's ISO/TC 184/SC 4/TF 1 meeting Keith agreed with the 2nd approach, which is to treat NOTEs, EXAMPLEs etc as separate Annotated EXPRESS tagged remarks. We will need to update the annotation generator.

**Comment by HassanAkbar** [comment]
@ronaldtse I've updated the annotation generator and now the annotated express file contains the following

```
(*"Activity_arm.Activity"
An **Activity** is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that **Activity** is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the **chosen_method** attribute.
*)



(*"Activity_arm.Activity.__example"
Change, distilling, design, a process to drill a hole, and a task such as training someone, are examples of activities.
*)


(*"Activity_arm.Activity.__note"
Status information identifying the level of completion of each activity may be provided within an instance of <<express:Activity_arm.Activity_status,Activity_status>>.
*)



(*"Activity_arm.Activity.__note"
The items that are affected by an *Activity*, for example as input or output, may be identified within an instance of <<express:Activity_arm.Applied_activity_assignment,Applied_activity_assignment>>.
*)
```

My question is do we need to include the `notes` and `examples` as `notes` and `examples` in the new definition?

**Comment by ronaldtse** [comment]
Yes, we need to translate those into:
```adoc
[NOTE]
--
...
--
```

and 
```adoc
[example]
====
...
====
```

**Comment by HassanAkbar** [comment]
Achieved in this PR

- As every Clause 4/5 term entry has a source, Removed the NOTE “incompletely defined in this document”.
- Updated `stepmod-annotate-all` to add note as `<entity>.__note` and example as `<entity>.__example` in annotated express files.
- Updated extract terms flow to read all the old definitions, notes, and examples from the `annotated_express` files and 
- Generated definition from entity name and template based on subtypes.
- Addd old definition as a `note` to new definition.
- Updated test cases accordingly.

File: spec/stepmod/terms_extractor_spec.rb [hunk 13]
```diff
@@ -39,46 +39,66 @@
...
 
+        [NOTE]
+        --
         An *action_directive* is an authoritative instrument that provides directions to achieve the specified results.
+        --
 
-
-        NOTE: This term is incompletely defined in this document.
-        Reference <<ISO_10303-41_2021>> for the complete definition.
+        [EXAMPLE]
```
Comment (ronaldtse) [inline spec/stepmod/terms_extractor_spec.rb:51]: We want lowercase `[example]` instead of `[EXAMPLE]`.

**Review by ronaldtse — CHANGES_REQUESTED** [review]
Looks good! Only one nit.

**Comment by ronaldtse** [comment]
Once this is merged we should run it on https://github.com/metanorma/iso-10303-2

File: spec/stepmod/terms_extractor_spec.rb (hunk 13 above)
Comment (ronaldtse) [reply spec/stepmod/terms_extractor_spec.rb]: Fixed.

**Review by ronaldtse — APPROVED** [review]

**Comment by ronaldtse** [comment]
We need a release bump @HassanAkbar .

**Comment by ronaldtse** [comment]
Actually this implementation is incomplete:
```
(*"action_schema.supported_item"
The **supported_item** allows for the designation of an <<express:action_schema.action_directive,action_directive>>, an <<express:action_schema.action,action>>, or an <<express:action_schema.action_method,action_method>>.
*)

(*".__note"
This specifies the use of an <<express:action_schema.action_resource,action_resource>>.
*)
```

Notice that the name of the `__note` is missing.

**Pull request merged**